### ISSUE TITLE
Load an ES2015 module config file 

### ISSUE THREAD

**Issue by danawoodman** [davidtheclark/cosmiconfig#109]
I am writing a CLI tool where I want my end users to be able to write their config file in ES6/7/Next, how can I accomplish this?

Eg, my `foo.config.js` file:
```js
import foo from './foo'

export default {  foo, someOtherOption: true }
```

**Comment by danawoodman** [comment]
I just implemented this in one of my own CLI tools and it should be as easy as importing `babel-register` and then importing the `default` export of its and ES module

**Comment by danawoodman** [comment]
Had to stop using cosmiconfig because my tool needs to support ES projects which may define variables in files using ES modules which are then included into the config file.

**Comment by davidtheclark** [comment]
Open to ideas here!

**Comment by danawoodman** [comment]
Basically I did this:

```js
import 'babel-register'

function importModule(module) {
  return module.__esModule ? module.default : module
}

const config = importModule(require(path.join(process.cwd(), 'myapp.config.js')))
```

**Comment by davidtheclark** [comment]
I changed the title of this issue to reflect that it is a new feature request. PRs welcome.

**Comment by chrisblossom** [comment]
This is how I support ``es modules`` with ``cosmiconfig`` (I think got the idea from eslint or babel). Maybe this should be handled by the end user because otherwise ``export default`` is only supported. If that is okay/expected, I could put in a PR to address this issue. Would require a major version bump because it would be a breaking change though.

```js
import cosmiconfig from 'cosmiconfig';

function transform(result) {
    let config = result.config;

    /**
     * Handle ES Modules
     */
    if (typeof config === 'object' && config.__esModule) {
        if (config.default) {
            config = config.default;
        } else {
            throw new Error(
                `${result.filepath} must use default export with ES Modules`,
            );
        }
    }

    return {
        ...result,
        config,
    };
}

const explorer = cosmiconfig('cfg', {
    rcExtensions: true,
    sync: true,
    transform,
});

function loadOptionsFile() {
    const { config } = explorer.load();

    return config;
}
```

**Comment by danawoodman** [comment]
What I mentioned above should be simpler and not require a major version bump AFAIK

**Comment by chrisblossom** [comment]
@danawoodman we solved it the same way, I just gave a full example to use with cosmiconfig's transform option. The reason both require a version bump is because if someone is using a non-default export it would break their code. If this is expected, then a major version bump isn't required.

**Comment by danawoodman** [comment]
Why would someone use a non-default export if using ES modules?

**Comment by chrisblossom** [comment]
I have no idea, just pointing out it has a chance to break someone's code.

I'm working on a PR for this.

**Comment by chrisblossom** [comment]
@davidtheclark @sudo-suhas Is the expectation that the user can directly load an ES Module config or is it up to the user to compile the config files into common js themselves? If it is to directly load ES Modules, babel will need to become a dependency. Is this wanted/acceptable?

**Comment by davidtheclark** [comment]
@chrisblossom I think supporting only a `default` export seems fine. 

If we introduce Babel *into* cosmiconfig as a dependency, I think we'd have to draw a stark boundary to make it clear we're only going to compile modules, not other ES2015+ syntax. (Your users can include whatever other ES2015+ syntax is available in your supported versions of Node.) 

Here are a few paths that seem possible to me:

- If there's a package like require-from-string that supports ES2015 modules, we could use that.
- If require-from-string errors because of a syntax error, try running a Babel transform on the string the require-from-string-ing the result.
- We could drop require-from-string, use `require` directly, instead; then rely on the user to use a babel-register hook to allow `require` to read `export`s. (I don't know if there are non-Babel equivalents for babel-register?) (See past issues about dropping require-from-string: https://github.com/davidtheclark/cosmiconfig/issues/21, https://github.com/davidtheclark/cosmiconfig/pull/51)
- Maybe standard-things/esm solves this problem? https://github.com/standard-things/esm? If we follow the "Enable esm with a CJS bridge" pattern.

**Comment by chrisblossom** [comment]
@davidtheclark looks like someone moved the changes to master for v5 to a branch and reverted the changes to master followed by a force push (makes very difficult for contributors/team members when this happens). Which branch would you like this PR placed?

**Comment by olsonpm** [comment]
I'll let davidtheclark answer what branch you should push to - but I would think v5 as that shouldn't take too long to push out.

As for the force push - he [mentioned it here](https://github.com/davidtheclark/cosmiconfig/pull/107#issuecomment-372137631) and seemed to just forget to tag you.  It was just to separate the v5 work from master

**Comment by shellscape** [comment]
I'd like to contribute a vote to _not_ bundling babel with comiconfig. `webpack-serve` is being integrated into some rather larger packages that have some prominent users that are frankly, a bit overly sensitive about package cost.

**Comment by danawoodman** [comment]
If `require-from-string` is used that dependency on `babel-register` would not come into play unless for some reason `require-from-string` would fail. 

Also, I assume it is a non-issue when loading a CommonJS module?

Are your consumers worried about the runtime issues of using `bagel-register` or just having `babel-register` in their `node_modules`?

**Comment by olsonpm** [comment]
I assume it's not just babel-register, but [its dependencies](https://www.npmjs.com/package/@babel/register)

>  core-js, find-cache-dir, home-or-tmp, lodash, mkdirp, pirates, source-map-support

**Comment by danawoodman** [comment]
What I'm trying to understand is the concern about file size in `node_modules` or runtime performance?

**Comment by chrisblossom** [comment]
@shellscape I agree that ``cosmiconfig`` should not carry a dependency to ``babel``. I also do not think it should pollute any globals, including ``require`` with ``babel-register``. This should be opt in by the consumer of the library.

I'm working on some solutions to this now. First step is replacing ``require-from-string`` with ``require`` as described earlier by @davidtheclark. This will allow the consuming library/app to use ``babel-register`` if desired.

**Comment by shellscape** [comment]
@danawoodman I didn't want to wade into the runtime issues of `babel-register` because they differ between v6 and v7 (beta) quite a bit, and are fraught with pitfalls (there are a metric poopton of issues with `babel-register` which is why they don't recommend using it in production). It's just really not a good idea.

The immediate concern is the total package cost, meaning the install of my module plus all of it's dependencies, and the dependencies dependencies and so on. Total disk size cost. It's silly and I don't like that people are so worried about it but whatevs.

**Comment by danawoodman** [comment]
Gotcha. So if @chrisblossom comes up with a non-babel option would the concern about disk size still be relevant?

**Comment by shellscape** [comment]
@danawoodman that's a negatory good buddy. non-babel would be preferred.

**Comment by davidtheclark** [comment]
Has anybody looked into https://github.com/standard-things/esm? I suggest we try that before any Babel business, because if it does the trick then our problems are solved, right?

**Comment by olsonpm** [comment]
i'll take a look at it tomorrow.  Been wanting a reason to give it a try.

**Comment by chrisblossom** [comment]
I'm fairly certain [standard-things/esm](https://github.com/standard-things/esm) pollutes ``require``, which would change how ``require`` works for anyone using this library. @jdalton could provide more information here.

**Comment by jdalton** [comment]
Hi @chrisblossom!

`esm` does not effect the global hook when used as part of a package. The entire reason it exists is to allow folks to ship ESM packages so that they could be consumed by CJS. 

#### index.js
```js
// Provide options as a parameter, environment variable, or rc file.
require = require("esm")(module/*, options*/)
module.exports = require("./main.js").default
```
In the example above `main.js` contains ESM code.

**Comment by jdalton** [comment]
If you have config files you can use it to load those too.
```js
const esmLoader = require("esm")(module)
const config = esmLoader(pathToConfig)
```

**Comment by jdalton** [comment]
Related Webpack v4 allows using `esm` to load webpack.config.js files as ESM using `webpack -r esm`.

**Comment by olsonpm** [comment]
Just an update - esm would be a convenient solution if I could figure out how to get it to compile a string directly (just like how require-from-string does it).  It will take me a bit to figure out a reasonable approach.

**Comment by chrisblossom** [comment]
@olsonpm I've got a PR to replace ``require-from-string`` with ``require`` almost ready. Just working out the last few issues. That should solve the issue you are having with esm.

**Comment by jdalton** [comment]
If you point me to a PR or WIP I can help iron anything out.

**Comment by chrisblossom** [comment]
@olsonpm [chrisblossom/cosmiconfig/tree/use-require](https://github.com/chrisblossom/cosmiconfig/tree/use-require) is my WIP for ``require-from-string``. Currently stuck on figuring out why Jest is intermittently failing on a ``successful-directories.test.js`` test. If anyone has any idea why this is happening only on the ``sync`` test, please let me know / put a PR in to fix.

**Comment by jdalton** [comment]
For compiling string I've made `esm` mimic the exposed Node ways to do so... I did find a typo though so I'll need to patch update to the `esm` package. After that though this will work:
```js
const requireESM = require("./index.js")(module)

const Module = requireESM("module")

const confMod = new Module("[config]", null)
confMod.filename = "fakePath"
confMod._compile("export { join } from 'path'", "fakePath")
confMod.loaded = true
console.log(confMod)
```

**Comment by olsonpm** [comment]
Ah, makes sense.  `requireESM('module')` is what I was missing.  Thank you

**Comment by olsonpm** [comment]
@davidtheclark - using the code above in lieu of require-from-string works great and doesn't require many changes.  [esm needs node >= 6](https://github.com/standard-things/esm/blob/master/package.json#L12-L14), so I was wondering whether you wanted to stick with supporting v4 and use [semver](https://github.com/npm/node-semver) to silently add the functionality - or if you wanted to bump the supported node version to v6.  I don't care either way, I just wanted to make sure before adding test cases for node 4 vs 6.

Also, one clarification I want to make is that my interpretation of this PR was to add support for `export default <config>` which is easy with esm.  It seems chrisblossom and possibly OP were hoping for babel transforms which means we'd have to use `require` instead of `readFile`.  If this is the direction then esm isn't necessary.

**Comment by jdalton** [comment]
RE: Node 6 requirement. It had Node 4 as the min bar but live bindings support for CJS required ES proxies. FWIW Node 4 EOLs at the end of April (so not too far away).

**Comment by chrisblossom** [comment]
Optimistically I'd like to see ``require`` used instead of ``require-from-string``, as well as an option that would allow the use of a custom function to handle javascript requires. If someone wants to use ``esm``, or ``babel`` they could easily with this option.

**Comment by davidtheclark** [comment]
**Use case 1: I want to allow my JS configs to use ES2015 modules, and whatever other ES2015+ syntax my the running version of Node allows for.**

It sounds like this will be fairly easy to provide users by using esm (whether we end up using require or require-from-string). So this seems to me like a relatively easy new feature to add, right? Any objections to adding it (independent of other use cases)?

**Use case 2: I want Babel powers, so not just ES2015 modules or ES2015+ syntax allowed by Node but also whatever else I feel like.**

I can't think of any way to do this well *without* depending on Babel. I've used up all my before-work-time this morning bouncing back and forth. I also think this use case is not very important. For a Node tool, having a hard requirement that your JS syntax needs to be supported by your version of Node — that's more than reasonable.

> an option that would allow the use of a custom function to handle loading javascript

This may work. Another possibility would be to add an option that makes JS results just return the file path, not try to load it — then you can do what you want with the file path. @chrisblossom or @danawoodman, any ideas whether one of these options would be better than the other?

**Comment by davidtheclark** [comment]
Regarding `require` vs require-from-string: I think we should switch to using `require` directly unless we come up with some solid reasons not to. 

The only one I think there was in the fast was that when you run across a JS config file, you end up having to look at it twice: once to see that it exists, then you `require` it and it gets read again. (As opposed to the other loaders, which try to read the file and parse it in one move.) But I don't think that's a serious performance, really, because the synchronous `require` call that duplicates a little bit of work only happens *once* and *only if* a JS config file is found.

So, to reiterate, if there aren't better reasons than that to use require-from-string, I think we should cut the dependency and use `require` directly.

**Comment by chrisblossom** [comment]
The reason to not use ``esm`` directly is one that @shellscape is worried about, increased bundle size. It also adds more complexity to this library, which could easily be added as an optional custom javascript loader. 

For use case 2, all we need to do is use ``require`` instead of ``require-from-string`` and the consuming library can use ``babel-register``. Also, with a custom loader function, it could run the found javascript file though babel (or typescript or w/e future compiler there is) and handle it that way.

I've thought about just returning a file path, but it would get more complicated because of [tryAllParsing](https://github.com/davidtheclark/cosmiconfig/blob/6256eb8e454e10e1b0c158a2e9bfd2dc984bf30d/src/loadDefinedFile.js#L70), [parseContent](https://github.com/davidtheclark/cosmiconfig/blob/6256eb8e454e10e1b0c158a2e9bfd2dc984bf30d/src/loadDefinedFile.js#L40) and [tryRequire](https://github.com/davidtheclark/cosmiconfig/blob/6256eb8e454e10e1b0c158a2e9bfd2dc984bf30d/src/loadDefinedFile.js#L92). How would ``cosmiconfig`` know what to return when the JS file has invalid JS syntax until compiled? I'm sure there are other examples, but this is what I can think of with my limited knowledge of the codebase.

TLDR/IMO, let the consuming library decide what non-current node features to support. We just need to make that possible.

**Comment by jdalton** [comment]
@chrisblossom I'm super interested on the how big is too big for a dependency. I used to gzip the loader so that it was ~75kb (zero dep). I have since stopped gzipping it since unzipping is a minor cost per execution, so its size on disk is a bit larger (though the over the wire size is similar since packages are downloaded as .tgz files).

**Comment by davidtheclark** [comment]
@chrisblossom: There was some concern voiced about babel-register above ("a metric poopton of issues ... which is why they don't recommend using it in production"). But I'm fine with offloading that onto the user: if they want to use babel-register, it's their choice.

If we were to use `require` directly to load JS files, which would allow the user to use a require-hook like babel-register if they want, what would be the use case for a user providing a custom JS-parsing function?

> The reason to not use esm directly is one that @shellscape is worried about, increased bundle size

That's an issue with Babel, definitely, but I really don't think it's an issue with esm — as @jdalton points out.

**Comment by chrisblossom** [comment]
@jdalton I personally don't care about the dependency size, just was pointing out why others might object. Although I do think it adds complexity that could be offloaded to the consumer if they want to support ES Modules via ``esm`` using a custom JS-parsing function. Also, if the user decides to use ``babel``, will ``esm`` have any conflicts/issues?

Use case for a custom JS-parsing function is so you wouldn't be forced to use ``babel-register``, because it does have negative issues (my use case, I don't want to use ``babel-register`` if at all possible, otherwise I'll need to figure out how to run ``cosmiconfig`` in a separate node ``spawn`` instance). Or they could support just es2015 modules with ``esm`` with a custom JS-parsing function. Or as said earlier, could use a ``typescript`` parser or some other parser that hasn't been released yet. The user could revert back to ``require-from-string`` if they wanted to as well with the custom parsing function.

I've read in other issues you want to keep this library as "low level" as possible, which I agree with. To me a custom JS-parsing function goes along with that idea.

**Comment by davidtheclark** [comment]
@chrisblossom With a custom parsing function, in each case the user would probably need to: read the file as a string, run the string through some compiler, then ... somehow load the string. Would the idea be that they'd transform the JS as a string and then cosmiconfig would require-from-string whatever string they return? Or they'd require-from-string it themselves? Or is there some other way to do this that people might use?

**Comment by chrisblossom** [comment]
The idea would be let the user completely dictate how javascript is loaded.

If the custom parser needs ``require-from-string`` they'd require that themselves.

The parser should only be passed the ``filePath`` and ``cosmiconfig``'s ``options``. Failed parsing should throw an error.


```js
// These are example parsers, they probably do not work exactly as they are.

function esmJSParser(filePath, options = {}) {
  const esmLoader = require('esm')(module);

  const config = esmLoader(filePath);

  return config;
}

function requireFromStringJSParser(filePath, options = {}) {
  const requireFromString = require('require-from-string');
  const fs = require('fs');

  const sync = !!options.sync;

  if (sync) {
    const fileAsString = fs.readFileSync(filePath, 'utf8');
    const config = requireFromString(fileAsString);

    return config;
  } else {
    // pretend fs.readFile returns a promise for this example
    return fs.readFile(filePath, 'utf8').then(contents => {
      const config = requireFromString(contents);
      return config;
    });
  }
}

function babelJSParser(filePath, options = {}) {
  const babel = require('babel-core');

  // I don't think this works. Just an example.
  // There is an async version that can be used with options.sync
  const config = babel.transformFileSync(filePath).code;

    if (!config) {
      throw new Error('js loading failed')
    }

  return config;
}
```

This would be called by a central JS require file, such as [chrisblossom/cosmiconfig/src/requireJs.js#L7](https://github.com/chrisblossom/cosmiconfig/blob/185ab02cecc3dd3b311d1f17042203ff37261dcf/src/requireJs.js#L7)

Does all of that make sense?

**Comment by shellscape** [comment]
Just throwing out a question: Could these different options be `optionalDependencies` and used if they've been installed? There'd have to be an order of precedence, and that might be a pain point, but for use-case under webpack-serve I'd either have to provide a pass-through option to cosmiconfig, or bundle one of the preferred parsers in my module. Neither of which is ideal. Probably no perfect solution here.

**Comment by davidtheclark** [comment]
@chrisblossom I think this mostly makes sense, except for a couple of points. 

- The last example, with Babel, is missing the step that actually loads the module from the transformed string. I wasn't sure if the sentence "This would be called by a central JS require file ..." was meant to apply to that example? If so, that function wouldn't actually be "central", right — if "central" means within cosmiconfig? It would be a function that the user implements, maybe using require-from-string. 
- "The parser should only be passed the filePath and cosmiconfig's options. " I don't think that the parser has any need for cosmiconfig's options. Was there some option you had in mind?

To clarify, then, here's what I understand as the plan:

- Add a `loadJs` option for creating an explorer, e.g.
  ```js
  const explorer = cosmiconfig('myModule', { loadJs: myJsLoader });
  ```
  `loadJs` would accept a `filePath` argument and would be expected to return either a Promise (async) or something else (sync). If it returns a Promise, cosmiconfig will wait for it to resolve and then assume that the value it resolves with is the config. If it returns anything else, cosmiconfig will assume that the value it returns is the config. If the function throws an error or returns a Promise that rejects, cosmiconfig will treat that like other parsing errors.
- If no `loadJs` option is provided (so, by default), cosmiconfig will `require` JS files directly. This allows users who want to to use a require hook.
- We should document examples of how to use esm to load ES2015 modules, or Babel to load whatever JS syntax you want.

This sounds like a fine plan to me 😬 

@shellscape The only method of loading JS that would be built in to cosmiconfig would be regular old `require`. So this plan includes no additional dependencies — in fact, we'd remove one (require-from-string). If you want ES2015 modules or anything else that's not supported by your version of Node, you'd include a `loadJs` function when you configure cosmiconfig. Does that sound good?

**Comment by shellscape** [comment]
@chrisblossom that'll require `webpack-serve` to provide a means to users, through the module, to interact with cosmiconfig. something more automatic that doesn't require `webpack-serve` to implement and support something special would be preferred. `webpack-dev-server` went down this path and over a few years it became bloated to the point of maintenance nightmare status. we're good with adding feature support provided by 3rd party modules, but we're desperately trying to avoid having to add features that we must support, in order to provide features from 3rd party modules. The users aren't direct consumers in this use-case -`webpack-serve` is acting as an intermediary.

That's why I suggested something along the lines of an `optionalDependency` which, if present, cosmiconfig would use. It seems like it could be a bit captain-redundancy-captain but perhaps a `.cosmiconfigrc` might provide end-users a means to specify that function without direct intervention by consumers like `webpack-serve`.

**Comment by davidtheclark** [comment]
@shellscape I don't understand this problem. webpack-serve does not need to change anything: it could just not accept ES2015 modules, as before, and things should continue working just the same. Or if you want to allow ES2015 modules, it would add a `parseJs` function, and that's that. Choosing whether the end user gets a special `parseJs` is just like choosing whether to allow JS configs at all (via the `js` option), or allow rc files with extensions (via the `rcExtensions` option). The consumer of cosmiconfig — in this case webpack-serve — makes the decision about what configurations are possible, then offers that to its users. You wouldn't need to allow the user to set `parseJs` anymore than you'd need to allow the user to set `rcExtensions`, `rcStrictJson`, or any other option.

**Comment by chrisblossom** [comment]
> The last example, with Babel, is missing the step that actually loads the module from the transformed string.

Yes, I think you are correct. I looked at the babel transform docs for 10 seconds to give a very limited example. The examples weren't necessarily supposed to work, just show what I envision a js parser function would look like / which arguments there would be. And that the required config would be returned.

> I wasn't sure if the sentence "This would be called by a central JS require file ..." was meant to apply to that example? If so, that function wouldn't actually be "central", right — if "central" means within cosmiconfig? It would be a function that the user implements, maybe using require-from-string.

Just meaning there would be one ``requireJs.js`` file, or whatever you'd call it, that would handle "requiring" all JS. Currently there is ``loadJs.js`` and ``require-from-string`` called in different places. Was not pertaining to the babel example.

> "The parser should only be passed the filePath and cosmiconfig's options. " I don't think that the parser has any need for cosmiconfig's options. Was there some option you had in mind?

The ``requireFromStringJSParser`` example showed one use case. The ``sync`` option, which I know is being changed but I think it should still be presented as a ``sync`` option to the parser. Unless we want accept ``async`` and ``sync`` ``parseJs`` functions. I'd prefer it as an option. The other one I know is ``ignoreEmpty``. I was thinking it would be better to just pass the complete ``options`` object so any future options would already be added / less future maintenance / less docs.

> loadJs would accept a filePath argument and would be expected to return either a Promise (async) or something else (sync). If it returns a Promise, cosmiconfig will wait for it to resolve and then assume that the value it resolves with is the config. If it returns anything else, cosmiconfig will assume that the value it returns is the config. If the function throws an error or returns a Promise that rejects, cosmiconfig will treat that like other parsing errors.

Sounds correct to me outside of I think the ``loadJs`` function also needs at least some ``options`` (see above). Also, I'd expect a thrown error to be shown to the user, instead of being overwritten by a generic failed to parse. At least with ``.js`` extensions. Maybe they already do?

> If no loadJs option is provided (so, by default), cosmiconfig will require JS files directly. This allows users who want to to use a require hook.

Yes, exactly what I am thinking.

> We should document examples of how to use esm to load ES2015 modules, or Babel to load whatever JS syntax you want.

Yes, and hopefully some of us can put together separate npm packages with them ready to go and possibly link them directly from the readme. I'll be looking into a ``babel`` one.

**Comment by sudo-suhas** [comment]
> If no `loadJs` option is provided (so, by default), cosmiconfig will `require` JS files directly.

Is `require` interchangeable with `require-from-string`? Wouldn't we need to use something like [`import-from`](https://github.com/sindresorhus/import-from)?

**Comment by davidtheclark** [comment]
@sudo-suhas Instead of reading the file as a string (which also verifies its existence) then passing that string to `require-from-string` to get the export, we can check the file's existence and then pass the filename to `require` to get the export.

> Wouldn't we need to use something like import-from?

We already have the absolute paths to the files that we'd end up `require`ing, so I don't think we have any need for `import-from`.

**Comment by shellscape** [comment]
Any additional traction we're not seeing listing in the issue on this one?

**Comment by chrisblossom** [comment]
@shellscape I've got a WIP branch, but have been holding off finishing it until https://github.com/davidtheclark/cosmiconfig/pull/120 has been completed.

**Comment by shellscape** [comment]
@davidtheclark thought I would share a hack we're testing as an interim solution. we're currently looking for potholes this hack might produce:

```js
const minimist = require('minimist');
const mock = require('mock-require');
const resolve = require('resolve').sync;

const cwd = process.cwd();
const argv = minimist(process.argv.slice(2));
const requires = [].concat(argv.require);

for (const module of requires) {
  const modulePath = resolve(module, { basedir: cwd });
  require(modulePath); // eslint-disable-line import/no-dynamic-require, global-require
}

mock('require-from-string', (...args) => {
  const [, path] = args;
  // eslint-disable-next-line import/no-dynamic-require, global-require
  return require(resolve(path, { basedir: cwd }));
});

const cosmiconfig = require('cosmiconfig');

const cosmicOptions = {
  rcExtensions: true,
  sync: true,
  js: 'ts.config.ts',
};
const explorer = cosmiconfig('ts', cosmicOptions);
const { config } = explorer.load() || {};

console.log('config', config); // eslint-disable-line
```

That script, run with `--require ts-node/register` works as expected. Removing the `js: ...` property from the config and using a `.js` file with ES6 modules, along with a `--require babel-register` flag works for ES6 modules.

There's probably some reasoning and history that I'm not aware of. But by basically taking `require-from-string` out of the loop here, we're able to load all kinds of goodies using their register/require-hook modules. One wonders, sans history of course, if `require-from-string` couldn't be removed altogether.

**Comment by davidtheclark** [comment]
@chrisblossom: The big refactoring PR is merged now.

> One wonders, sans history of course, if require-from-string couldn't be removed altogether.

@shellscape: The plan is in fact to remove it altogether. I think the `loadJs` option, in combination with a `jsExtensions` option (cf. https://github.com/davidtheclark/cosmiconfig/issues/122), might enable your use case.

**Comment by chrisblossom** [comment]
@davidtheclark Just an update...I've got this 95% ready to put in a PR to be reviewed. Just working out some flow issues. Should have it done in the next couple days.

**Comment by davidtheclark** [comment]
The v5 branch now makes this possible in a variety of ways — whether you want to write a custom loader function or use require hooks.
